tutorials/haskell/generics/tutorial.md

```diff
@@ -0,0 +1,584 @@
+       (C1 ('MetaCons "Nil" 'PrefixI 'False) U1
+        :+: C1
+              ('MetaCons "Cons" 'PrefixI 'False)
+              (S1
+                 ('MetaSel
+                    'Nothing 'NoSourceUnpackedness 'NoSourceStrictness 'DecidedLazy)
+                 (Rec0 a)
+               :*: S1
+                     ('MetaSel
+                        'Nothing 'NoSourceUnpackedness 'NoSourceStrictness 'DecidedLazy)
+                     (Rec0 (List a))))
+       p
+```
+
+Ah OK, there is just a little bit more to this representation thing…
+
+## Metadata wrappers
+
+Representation also has associated metadata. It may look messy and difficult
+to read, but I'll explain the logic behind it in a moment, it should help to
+clarify why it's done like this.
+
+First of all, metadata should not get in our way if we don't want to deal
+with it. Thus all metadata is attached using the same simple wrapper:
+
+```haskell
+newtype M1 i c f p = M1 { unM1 :: f p } -- ‘f p’ is what lives inside, U1 for example
+--         ^ ^
+--         | |
+--         | +---- compiler-generated data type that allows to get meta information
+--         |
+--  type-level tag, see below
+```
+
+The `i` type-level tag can be one of the three:
+
+* `D` for data type metadata `type D1 = M1 D`
+* `C` for constructor metadata `type C1 = M1 C`
+* `S` for record selector metadata `type S1 = M1 S`
+
+It should be understandable why metadata is attached this way. If we want
+to, we can ignore metadata easily, since it's just a wrapper:
+
+```haskell
+f (M1 x) = f x
+```
+
+If we want to look at a particular type of metadata we can specify the `i`
+type-level tag, or we can leave it unspecified to deal with all metadata at
+once.
+
+The `c` type is auto-generated by the compiler and encodes metadata on type
+level. Why not store the metadata on value level, in `M1` constructor? Well,
+if it were there, we would have to provide it if we wanted to generate some
+values, and providing metadata for already existing data type is certainly
+something that only compiler can do properly. With the current approach,
+given data type determines type of its representation, including metadata,
+so we don't have to bother.
+
+The last thing to mention about metadata is which wrappers are generated for
+you. They are very predictable:
+
+* The entire representation is wrapped in `D1`, which provides
+  datatype-level information (datatype name, module name, and whether it's a
+  `newtype`).
+
+* Every constructor is wrapped in `C1`, which provides information about
+  constructor (such as constructor name, fixity, and whether it's a record).
+
+* Every argument of constructor is wrapped with `S1` (even if it's not
+  actually a record selector), which tells us selector name.
+
+Look
+[at the Haddock](https://hackage.haskell.org/package/base-4.8.2.0/docs/GHC-Generics.html#t:Datatype),
+to find out names of functions that help extract the metadata. Using them is
```

Comment:
Wrappers are part of the representation. As we have seen wrapping in `M1` or `L1` adds context to whatever is inside of it.